hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/upgrade/UpgradeChecker.java

```diff
@@ -0,0 +1,143 @@
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.hadoop.ozone.OzoneConsts.CONTAINER_DB_NAME;
+
+/**
+ * This is the handler that process container upgrade checker.
+ */
+public class UpgradeChecker {
+
+  /*
+   * Verify that the datanode is in the shutdown state.
+   */
+  public Pair<Boolean, String> checkDatanodeNotStarted() {
+    String command =
+        "ps aux | grep org.apache.hadoop.ozone.HddsDatanodeService " +
+            "| grep -v grep";
+    try {
+      Process exec = Runtime.getRuntime().exec(command);
+      exec.waitFor(10, TimeUnit.SECONDS);
+      if (exec.exitValue() != 0 && exec.exitValue() != 1) {
+
+        return Pair.of(false,
+            String.format("Return code of the command %s was %d", command,
+                exec.exitValue()));
+      } else {
+        final boolean dnStarted =
+            IOUtils.toString(exec.getInputStream(), StandardCharsets.UTF_8)
+                .contains("proc_datanode");
+        if (dnStarted) {
+          return Pair.of(false,
+              "This command needs to be executed with datanode turned off.");
+        }
+      }
+    } catch (IOException | InterruptedException e) {
+
+      return Pair.of(false,
+          String.format("Run the command %s has error", command));
+
+    }
+    return Pair.of(true, "");
+  }
+
+  public Pair<Integer, Integer> getLayoutVersion(OzoneConfiguration conf)
+      throws IOException {
+    SCMStorageConfig scmStorageConfig = new SCMStorageConfig(conf);
+    HDDSLayoutVersionManager scmLayoutVersionManager =
+        new HDDSLayoutVersionManager(
+            scmStorageConfig.getLayoutVersion());
+    final int metadataLayoutVersion =
+        scmLayoutVersionManager.getMetadataLayoutVersion();
+    final int softwareLayoutVersion =
+        scmLayoutVersionManager.getSoftwareLayoutVersion();
+
+    return Pair.of(softwareLayoutVersion, metadataLayoutVersion);
+  }
+
+  public List<HddsVolume> getAllVolume(DatanodeDetails detail,
+      OzoneConfiguration configuration) throws IOException {
+    final MutableVolumeSet dataVolumeSet = UpgradeUtils
+        .getHddsVolumes(configuration, StorageVolume.VolumeType.DATA_VOLUME,
+            detail.getUuidString());
+    return StorageVolumeUtil.getHddsVolumesList(dataVolumeSet.getVolumesList());
+  }
+
+  public List<File> getVolumeDBPath(List<HddsVolume> hddsVolumes) {
+    List<File> dbPaths = new ArrayList<>();
+
+    for (HddsVolume storageVolume : hddsVolumes) {
+      File clusterIdDir =
+          new File(storageVolume.getStorageDir(), storageVolume.getClusterID());
+      File storageIdDir =
+          new File(clusterIdDir, storageVolume.getStorageID());
+      final File containerDBPath =
+          new File(storageIdDir, CONTAINER_DB_NAME);
```

Comment:
Use UpgradeUtils#getContainerDBPath here.

Reply:
@guohao-rosicky , this get container DB path logic can be replaced using UpgradeUtils#getContainerDBPath.